Analyze/report performance test results

zyrolasting opened the pull request:
This code reads performance test results, and can optionally present a tabular summary. This functionality is not yet integrated into a CI system, so I need a review over how I'm extracting data (Empirical means are necessary due to  #108)

e.g.

```
╔════════╤════════════╤════════════╗
║        │ topn       │ wf         ║
╟────────┼────────────┼────────────╢
║ abd611 │ 473277.432 │ 179270.421 ║
╚════════╧════════════╧════════════╝
```

(The borders come from NPM's [table][] package, and are easily removed)

[table]: https://www.npmjs.com/package/table
 
To reproduce the above table, do this:

```
const { analyze, keepWidth, keepTimed, render } = require('./perf-analysis.js');

// change me
const resultsPath = '/path/to/pash/evaluation/results';

const data = analyze(resultsPath);

// Keep only the timed results given --width 64
const filtered = keepTimed(keepWidth(data, 64)).filter(({variant}) => variant === 'distr');

// render() takes the data in an array like [['commit-hash', { ... }], ['commit-hash', { ... }]]
console.log(render([['abd611', filtered]], ['wf', 'topn']))
```

EDIT: Updated filter in example to show only `_distr.time` timings.

zyrolasting commented on scripts/remote/controller/perf-analysis.js:
```diff
@@ -0,0 +1,93 @@
+      parseInt((path.basename(p, path.extname(p)).match(/_(\d+)_/) || [])[1]);
+
+const extractVariant = (p) =>
+      (p.match(/_distr_?([^\.]+)/) || [])[1];
```
I can't issue a change request on my own PR, so I'll just flag this here.

This pattern looks for test variants that strictly follow `_distr_?`, which means `_seq` is not seen as a test variant. It might be better to parse from right to left as @angelhof mentioned in #108.

angelhof reviewed (approved):
This looks good to me! One note worth considering is that `_distr_auto_split.time` are the _best_ PaSh results. When they don't exist the reason is usually that `_distr.time` corresponds with `_distr_auto_split.time`. However I don't think that this should be a concern for now.

> so I need a review over how I'm extracting data

What exactly is the issue now with data extraction?

zyrolasting commented:
> What exactly is the issue now with data extraction?

It's just that I'm yanking out values from text logs using regular expressions. That, and you were mentioning some of the results are old. Altogether the process felt error prone, so I wanted to hear if the values you can produce using this code are consistent with your expectations.

angelhof commented:
OK got it! 

There are a couple points that need to be addressed to get the table results that we actually want.
- There could be multiple "Execution time:" lines in a log file, so we would want to sum them to get correct results. Alternatively, we could also parse the total time (produced from the time command) which includes all of the PaSh overhead (compilation, optimization, parsing, etc). Maybe we want both :)
- We would usually go for speedup. To show that we would need to get the `*seq.time` time (which is produced by `time` and not PaSh so it has to be parsed differently) and divide it by the PaSh execution time. Then we can report the sequential time next to the speedup (e.g. seq-time: 100s (x10 pash-speedup))
- For the most up to date results (that we would want to visualize in the table), we would need to look at the perf-benchmarks that are run now. I think that you are running `evaluation/eurosys/execute_eurosys_one_liners.sh -m` and `evaluation/eurosys/execute_unix_benchmarks.sh -s`. The results for the first are stored in `evaluation/results/eurosys_small/` directory and the files to look for PaSh are the `distr_auto_split.time` files. The results for the second are stored in `evaluation/results/unix50_4_1073741824` again in the `distr_auto_split.time` files.
- For now, I would fix a width (e.g. 16) for each benchmark set, and the variant, and just visualize them in a commit by commit table as follows (the numbers are made up):

```
Results for width 16, full PaSh (distr_auto_split), eurosys-one-liners:
╔════════╤═════════════╤═════════════╗
║        │ topn        │ wf          ║
╟────────┼─────────────┼─────────────╢
║ abd611 │ 473s, x10.3 │ 270s, x5.4  ║
╚════════╧═════════════╧═════════════╝
```

zyrolasting commented:
@angelhof Thank you for the excellent and detailed feedback! I'm already on the way with last two bullets you mentioned after convos with @nvasilakis, so we're good there. Just a couple of follow up questions:

> Summing execution time

Using vanilla JS I'd have to parse these as floats and add them together. Is floating point imprecision tolerable here? Would I need to report an error margin alongside the tables?
 
> Maybe we want [multiple times]

Is that something that has to be implemented in this PR?

angelhof commented:
Great!

> Using vanilla JS I'd have to parse these as floats and add them together. Is floating point imprecision tolerable here? Would I need to report an error margin alongside the tables?

 I think floating point is fine (that is what I have been using until now too).

> Is that something that has to be implemented in this PR?

I don't think so, I just mentioned it for more context and as a possible extension later :)

zyrolasting commented:
@angelhof I updated the code to handle the requirements you mentioned. I needed this branch for future work anyway, so please forgive my asking for a second review pass. 

* Added README
* Includes a utility module (`history.js`) I'll need to integrate this into CI later
* Now reporting the sum of all PaSh execution times in a file.
* I figured out a way to report the data contextually. The below screenshots show that timing information is presented based on available data.
    - Sequential tests: `real`, `user`, `sys` timings.
    - Dist. tests + variants, Runtime in seconds, with speedup factor if adjacent `_seq.time` is available 

(*NOTE:* I intentionally left out the commit column here. It's trivial to add back once I have that data ready.)

![image](https://user-images.githubusercontent.com/1312121/106672858-c9bfb180-657e-11eb-980d-fb6dc855ca0f.png)
![image](https://user-images.githubusercontent.com/1312121/106673193-418ddc00-657f-11eb-939d-047d40d117b8.png)

angelhof commented:
> @angelhof I updated the code to handle the requirements you mentioned. I needed this branch for future work anyway, so please forgive my asking for a second review pass.

No worries! Thanks a lot for the changes :)

Everything looks good for now. I think we should merge after Nikos reviews.

At some later point we can discuss about presentation details and inconsistencies regarding the timing data that is produced by PaSh and hopefully clean up and rebuild the experiment running infrastructure to have consistent folder/file names.

nvasilakis commented:
@zyrolasting Thanks for all your work here!

What is the purpose of `history.js`? Is it manipulate results from all commits as a little JSON DB?

More generally, a simple solution here would be to copy the behavior from the "correctness" reports, but just change the data you're using and their presentation: the same way we have multiple columns there with different _correctness_ numbers, we would simply have different columns here with different _performance_ numbers. Summaries would be flat append-only logs (and optionally per-commit files would have more details about the performance result of a single commit — but this can come later).

zyrolasting commented:
> What is the purpose of history.js? Is it manipulate results from all commits as a little JSON DB?

In the context of performance tests, yes. I mentioned this in my other comment.
